cli/flox-rust-sdk/src/models/lockfile.rs

```diff
@@ -610,7 +739,222 @@ impl LockedManifestCatalog {
+                            if level != MessageLevel::Error {
+                                tracing::debug!(
+                                    level = level.to_string(),
+                                    msg = inner.msg,
+                                    "non-error resolution message"
+                                );
+                                continue;
+                            }
+                        }
+                        // If we don't have a level, I guess we have to treat it like an error
+                        tracing::debug!("pushing fallback message");
+                        let failure = ResolutionFailure::ConstraintsTooTight {
+                            fallback_msg: inner.msg.clone(),
+                            group: group.name.clone(),
+                        };
+                        failures.push(failure);
+                    },
+                }
+            }
+        }
+        failures
+    }
+
+    /// Converts a "attr path not found" message into different resolution failures
+    fn attr_path_not_found_failure(
+        r_msg: &MsgAttrPathNotFound,
+        manifest: &TypedManifestCatalog,
+    ) -> Option<ResolutionFailure> {
+        // If we have a level and it's not an error, skip this message
+        if let Some(level) = r_msg.level {
+            if level != MessageLevel::Error {
+                tracing::debug!(
+                    level = level.to_string(),
+                    msg = r_msg.msg,
+                    "non-error resolution message"
+                );
+                return None;
+            }
+        }
+        if r_msg.valid_systems.is_empty() {
+            tracing::debug!(
+                attr_path = r_msg.attr_path,
+                "no valid systems for requested attr_path"
+            );
+            Some(ResolutionFailure::PackageNotFound {
+                install_id: r_msg.install_id.clone(),
+                attr_path: r_msg.attr_path.clone(),
+            })
+        } else {
+            let (valid_systems, requested_systems) =
+                Self::determine_valid_and_requested_systems(r_msg, manifest)?;
+            let mut diff = requested_systems
+                .difference(&valid_systems)
+                .cloned()
+                .collect::<Vec<_>>();
+            diff.sort();
+            let mut available = r_msg.valid_systems.clone();
+            available.sort();
+            Some(ResolutionFailure::PackageUnavailableOnSomeSystems {
+                install_id: r_msg.install_id.clone(),
+                attr_path: r_msg.attr_path.clone(),
+                invalid_systems: diff,
+                valid_systems: available,
+            })
+        }
+    }
+
+    /// Determines which systems a package is available for and which it was requested on
+    fn determine_valid_and_requested_systems(
+        r_msg: &MsgAttrPathNotFound,
+        manifest: &TypedManifestCatalog,
+    ) -> Option<(HashSet<System>, HashSet<System>)> {
```

Comment:
Question: are these used in messages? might be worth using btreesets to ensure the message does not change due to random ordering of the elements

Reply:
apparently we sort these after diffing: https://github.com/flox/flox/blob/zmitchell/1482-better-resolution-messages/cli/flox-rust-sdk/src/models/lockfile.rs#L857